quinn-proto/src/connection/qlog.rs

```diff
@@ -281,6 +285,111 @@ impl QlogSink {
+        };
+
+        let timer_type: Option<TimerType> = match timer {
+            Timer::Conn(conn_timer) => match conn_timer {
+                ConnTimer::Idle => Some(QlogTimerType::IdleTimeout.into()),
+                ConnTimer::Close => Some(TimerType::custom("close")),
+                ConnTimer::KeyDiscard => Some(TimerType::custom("key_discard")),
+                ConnTimer::KeepAlive => None,
+                ConnTimer::PushNewCid => Some(TimerType::custom("push_new_cid")),
+            },
+            Timer::PerPath(_, path_timer) => match path_timer {
+                PathTimer::LossDetection => Some(QlogTimerType::LossTimeout.into()),
+                PathTimer::PathIdle => Some(TimerType::custom("path_idle")),
+                PathTimer::PathValidation => Some(QlogTimerType::PathValidation.into()),
+                PathTimer::PathChallengeLost => Some(TimerType::custom("path_challenge_lost")),
+                PathTimer::PathOpen => Some(TimerType::custom("path_open")),
+                PathTimer::PathKeepAlive => None,
+                PathTimer::Pacing => None,
+                PathTimer::MaxAckDelay => Some(QlogTimerType::Ack.into()),
+                PathTimer::PathAbandoned => Some(TimerType::custom("path_abandoned")),
+                PathTimer::PathNotAbandoned => Some(TimerType::custom("path_not_abandoned")),
+            },
+        };
+
+        let Some(timer_type) = timer_type else {
+            return;
+        };
+
+        let delta = match op {
+            TimerOp::Set(instant) => instant
+                .checked_duration_since(now)
+                .map(|dur| dur.as_secs_f32() * 1000.),
+            _ => None,
+        };
+        let path_id = match timer {
+            Timer::Conn(_) => None,
+            Timer::PerPath(path_id, _) => Some(path_id.as_u32() as u64),
+        };
+
+        let event_type = match op {
+            TimerOp::Set(_) => TimerEventType::Set,
+            TimerOp::Expire => TimerEventType::Expired,
+            TimerOp::Cancelled => TimerEventType::Cancelled,
+        };
+
+        let event = TimerUpdated {
+            path_id,
+            timer_type: Some(timer_type),
+            timer_id: None,
+            packet_number_space: None,
+            event_type,
+            delta,
+        };
+        stream.emit_event(EventData::TimerUpdated(event), now);
+    }
+
+    /// Returns a [`QlogSinkWithTime`] that passes along a `now` timestamp.
+    ///
+    /// This may be used if you want to pass a [`QlogSink`] downwards together with the current
+    /// `now` timestamp, to not have to pass the latter separately as an additional argument just
+    /// for qlog support.
+    pub(super) fn with_time(&self, now: Instant) -> QlogSinkWithTime<'_> {
+        QlogSinkWithTime { sink: self, now }
+    }
+}
+
+/// A [`QlogSink`] with a `now` timestamp.
+#[cfg_attr(not(feature = "qlog"), allow(unused))]
+pub(super) struct QlogSinkWithTime<'a> {
+    sink: &'a QlogSink,
+    now: Instant,
+}
```

Comment:
Should the fields not be cfg'd out so that it becomes a zero-sized struct when the qlog feature is disabled?

Reply:
Right, can do that.